src/porepy/numerics/time_step_control.py

```diff
@@ -170,120 +366,173 @@ def __init__(
                 print(f"Relaxing time step. Next dt = {self.dt}.")
-        elif iterations >= self.iter_upp:
-            self.dt = self.dt * self.iter_upp_factor
+        elif iterations >= self.iter_optimal_range[1]:  # (C2)
+            self.dt = self.dt * self.iter_relax_factors[0]
             if self._print_info:
                 print(f"Restricting time step. Next dt = {self.dt}.")
+        else:
+            pass  # (C3)
+
+    def _adaptation_based_on_recomputation(self) -> None:
+        """Adapt (decrease) time step when the solution failed to converge.
+
+        Raises: ValueError if recomp_attemps > max_recomp_attempts. That is, when the maximum
+            number of recomputation attempts has been exhausted.
 
-        # Check if the calculated time step is less than the minimum allowable time step
-        if self.dt < self.dt_min:
-            self.dt = self.dt_min
+        """
+
+        if self._recomp_num < self.recomp_max:
+            # If the solution did not converge AND we are allowed to recompute it, then:
+            #   (S1) Update simulation time since solution will be recomputed.
+            #   (S2) Decrease time step multiplying it by the recomputing factor < 1.
+            #   (S3) Increase counter that keeps track of the number of times the solution
+            #        was recomputed.
+
+            # Note that iterations is not really used here. So, as long as
+            # recompute_solution = True and recomputation_attempts < max_recomp_attempts,
+            # the method is entirely agnostic to the number of iterations passed. This
+            # design choice was made to give more flexibility, in the sense that we are
+            # not limiting the recomputation criteria to _only_ reaching the maximum
+            # number of iterations, even though that is the primary intended usage.
+            self.time -= self.dt  # (S1)
+            self.dt *= self.recomp_factor  # (S2)
+            self._recomp_num += 1  # (S3)
             if self._print_info:
-                print(f"Calculated dt < dt_min. Using dt_min = {self.dt_min} instead.")
+                msg = (
+                    "Solution did not converge and will be recomputed."
+                    f" Recomputing attempt #{self._recomp_num}. Next dt = {self.dt}."
+                )
+                print(msg)
+        else:
+            # The solution did not converge AND we exhausted all recomputation attempts
+            msg = (
+                f"Solution did not converge after {self.recomp_max} recomputing "
+                "attempts."
+            )
+            raise ValueError(msg)
 
-        # Check if the calculated time step is greater than the maximum allowable time step
-        if self.dt > self.dt_max:
-            self.dt = self.dt_max
+    def _correction_based_on_dt_min(self) -> None:
+        """Correct time step if dt < dt_min."""
+        if self.dt < self.dt_min_max[0]:
+            self.dt = self.dt_min_max[0]
```

Comment:
Indeed, this is a very good idea. In fact, it will save the user several hopeless iteration attempts.

I will, however, incorporate a different error message for this case, just to make debugging easier for the user.

Reply:
I included this check, with a slight different logic, but with same result. Basically, instead of manually setting self._recomp_num = self.recomp_max, I've added an extra check where an error is raised if adaptation based on recomputation is attempted when dt = dt_min.